_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2045 @@
+  source code itself. This can further decrease the effect of a potential breach of the CI server, as the attacker will
+  not have the ability to read or modify the infrastructure code to use the pipeline to their advantage.
+
+These mitigations alone will not prevent all attacks defined in the threat model. For example, an internal
+attacker with access to the source code can still do damage to the target environments by merging in code that removes
+all the infrastructure resources, thereby destroying all infrastructure when the `apply` command is run. Or, they could
+expose secrets by writing infrastructure code that will leak the secrets in the logs via a `local-exec` provisioner.
+Note that __any__ CI/CD solution can likely be compromised if an attacker has full access to your source code.
+
+For these types of threats, your best bet is to implement various policies and controls on the source control repository
+and build configurations:
+
+<<use_approval_flows>>::
+  In addition to providing a moment to pause and inspect the exact infrastructure changes that are about to be deployed,
+  approval workflows in the CI server can mitigate attacks such that attackers will need enough privileges on the CI
+  server to approve builds in order to actually modify infrastructure. This can mitigate potential attacks where the
+  attacker has access to the CI server to trigger arbitrary builds manually (e.g to run a previous job that is deplying
+  an older version to roll back the infrastructure), but not enough access to approve the job. Note that this will not
+  mitigate potential threats from internal attackers who have enough permissions to approve builds.
+
+<<lock_down_vcs_systems>>::
+  As mentioned in the previous section, it is important that you implement various controls on the VCS repositories.
+  Once you implement a CI/CD pipeline, access to source code translates to access to your infrastructure environments,
+  so you will want to reflect the same kind of security controls you implement on your environments in your VCS
+  repositories.
+
+Avoid logging secrets::
+  Our threat model assumes that attackers can get access to the CI servers, which means they will have access to the
+  deployment logs. This will include detailed outputs from a `terraform plan` or `apply`. While it is impossible to
+  prevent terraform from leaking secrets into the state, it is possible to avoid terraform from logging sensitive
+  information. Make use of pgp encryption functions or encrypted environment variables / config files (in the case of
+  service deployments) to ensure sensitive data does not show up in the plan output. Additionally, tag sensitive outputs
+  with the `sensitive` keyword so that terraform will mask the outputs.
+
+
+[[deployment_walkthrough]]
+== Deployment walkthrough
+
+Let’s now walk through the step-by-step process of how to create a production-grade CI/CD pipeline for your
```

Comment:
This should be defined explicitly as a standalone item in the production design section.

Reply:
That was what the `serverless` section was about. Otherwise, there really is no difference between this system and TFE, so I couldn't flat out say no TFE.

Let me see if I can reframe it better...

Reply:
See 427a56bc1a9eb400331fdb34cf82e44a9d6aabcb for reframe